proposals/Exceptions.md

```diff
@@ -143,42 +93,85 @@ Exception indices are used by:
-exception. The representation of an exception is left to the implementation.
-
-### Try and catch blocks
+### Try-catch blocks
 
 A _try block_ defines a list of instructions that may need to process exceptions
 and/or clean up state when an exception is thrown. Like other higher-level
 constructs, a try block begins with a `try` instruction, and ends with an `end`
-instruction. That is, a try block is sequence of instructions having the
+instruction. That is, a try-catch block is sequence of instructions having the
 following form:
 
 ```
 try blocktype
   instruction*
-catch
+catch i
+  instruction*
+catch j
+  instruction*
+...
+catch_all
   instruction*
 end
 ```
 
-A try block ends with a `catch block` that is defined by the list of
-instructions after the `catch` instruction.
+A try-catch block contains zero or more `catch` blocks and zero or one
+`catch_all` block. All `catch` blocks should precede the `catch_all` block, if
+any. The `catch`/`catch_all` instructions (within the try construct) are called
+the _catching_ instructions. There should be at least `catch` or `catch_all`
+block within a try-catch block.
+
+The _body_ of the try block is the list of instructions before the first
+catching instruction. The _body_ of each catch block is the sequence of
+instructions following the corresponding catching instruction before the next
+catching instruction (or the `end` instruction if it is the last catching
+block).
+
+The `catch` instruction has an exception tag associated with it. The tag
+identifies what exceptions it can catch. That is, any exception created with the
+corresponding exception tag. Catch blocks that begin with a `catch` instruction
+are considered _tagged_ catch blocks.
+
+The last catching instruction of a try-catch block can be the `catch_all`
+instruction. If it begins with the `catch_all` instruction, it defines the
+_default_ catch block. The default catch block has no exception type, and is
+used to catch all exceptions not caught by any of the tagged catch blocks. The
+term 'catch block' refers to both `catch` and `catch_all` blocks.
 
 Try blocks, like control-flow blocks, have a _block type_. The block type of a
 try block defines the values yielded by evaluating the try block when either no
 exception is thrown, or the exception is successfully caught by the catch block.
 Because `try` and `end` instructions define a control-flow block, they can be
 targets for branches (`br` and `br_if`) as well.
 
-In the initial implementation, try blocks may only yield 0 or 1 values.
+### Try-unwind blocks
+
+Try blocks can also be used with the `unwind` instruction. A try-unwind block
+contains an `unwind` block with the following form:
+
+```
+try blocktype
+  instruction*
+unwind
+  instruction*
+end
+```
+
+The `unwind` block is meant to contain cleanup instructions, such as
+destructors, in case any instruction in the corresponding try block throws.
```

Comment:
What are these "destructors" that the `unwind` block runs? Is this a special category of Wasm instructions?

Reply:
Destructors here are from C++. They are C++'s analog of `finally`. However, their semantics slightly differ across C++ versions. In older versions, if a destructor throws an exception during "normal" clean up, then the exception is propagated, but if a destructor throws an exception during "excpetional" clean up, i.e. unwinding caused by an exception being thrown, then the program is supposed to terminate.

Reply:
@ioannad `unwind` blocks contain routines for cleanups that need to run when the stack is unwound. In C++ that's mostly destructors, and `finally` can belong here in compilers where normal-pass `finally` and exception-pass `finally` are compiled separately. But we don't have support for any languages that have `finally` construct yet. I'll change this term to "C++ destructors" to make the example clearer.

Reply:
@aheejin I suppose here is the right place to make [this comment](https://github.com/WebAssembly/exception-handling/issues/87#issuecomment-707683679). 

I am not sure I understand how one can embed C++ instructions (destructors) in a WebAssembly code block, or how we can specify a WebAssembly instruction as containing instructions from another language (C++).

Reply:
I believe the point of mentioning C++ destructors is to describe the intended use case of this feature (and its semantics by rough analogy to that use case), not to actually involve C++ in the formal semantics. "Destructors" are not and should not become a Wasm concept in this proposal.

Reply:
@ioannad Destructors are just function calls. Like,
```cpp
void test() {
  Object o;
  ... do some work ...
}
```
In this code, the compiler will automatically generate calls like `call ~Object()` (the real function name will be some mangled name) at the end of this function. So they are just plain function calls that need to run as cleanup whenever we exit the scope. This destructor should run both when exit the scope normally and exceptionally, but many compilers, such as LLVM, compile those two path separately, because the two paths have some different things to handle.

Anyway, this was used as an example, but if people think this term is confusing, I can remove it.

By the way, I was maybe not doing a very good job on explaining this when we were having a meeting. Did you think `unwind` is the same as `finally` then?

Reply:
It seems I have forgotten to answer this comment, apologies @aheejin !

> Anyway, this was used as an example, but if people think this term is confusing, I can remove it.

To me it *was* confusing because I am thinking of the Wasm spec as self-standing, and I don't really think about translations from C++, as I have not worked on any tooling, just on the SpiderMonkey implementation. This might be just me though, so I wouldn't mind either way now that I understand the intention behind it. You pushed a commit which removes this anyway, which I think is fine.

> By the way, I was maybe not doing a very good job on explaining this when we were having a meeting.

I don't think we talked about unwind much, so apologies for not having asked more questions there.

> Did you think unwind is the same as finally then?

In a way I did; I thought unwind is the same as finally when an exception is thrown, this is why I [initially wrote a formal spec overview](https://github.com/WebAssembly/exception-handling/issues/87#issuecomment-702323595) with that
behaviour for unwind. When I understood that this is equivalent to `catch_all ... rethrow 0 end` (modulo the actual behaviour of unwind's `rethrow` in a future 2-phase-unwinding proposal), I simplified the spec to the one in #143 by "hardcoding" this equivalence for now.